src/metabase/lib/filter/update.cljc

```diff
@@ -0,0 +1,179 @@
+  [query        :- ::lib.schema/query
+   stage-number :- :int
+   column       :- ::lib.schema.metadata/column]
+  (reduce
+   (fn [query [_tag _opts expr :as filter-clause]]
+     (if (is-ref-for-column? expr column)
+       (lib.remove-replace/remove-clause query stage-number filter-clause)
+       query))
+   query
+   (lib.filter/filters query stage-number)))
+
+(mu/defn update-numeric-filter :- ::lib.schema/query
+  "Add or update a filter against `numeric-column`. Adapted from
+  https://github.com/metabase/metabase/blob/98bcd7fc3102bd7c07e8b68878c3738f3cb8727b/frontend/src/metabase-lib/queries/utils/actions.js#L151-L154"
+  ([query numeric-column start end]
+   (update-numeric-filter query -1 numeric-column start end))
+
+  ([query          :- ::lib.schema/query
+    stage-number   :- :int
+    numeric-column :- ::lib.schema.metadata/column
+    start          :- number?
+    end            :- number?]
+   (let [[start end] (sort [start end])]
+     (-> query
+         (remove-existing-filters-against-column stage-number numeric-column)
+         (lib.filter/filter stage-number (lib.filter/between numeric-column start end))))))
+
+(def ^:private temporal-filter-min-num-rows
+  "Minimum number of rows an updated query should return before; if it will return less than this, switch to
+  the [[unit->next-largest-unit]]. E.g. if we zoom in on a query using unit is `:day` and the zoomed in query would
+  only return 2 rows, switch the unit to `:minute`."
+  4)
+
+(def ^:private unit->next-largest-unit
+  "E.g. the next largest unit after `:hour` is `:minute`."
+  (let [units [:minute :hour :day :week :month :quarter :year]]
+    (zipmap units (cons nil units))))
+
+(mu/defn ^:private temporal-filter-find-best-breakout-unit :- ::lib.schema.temporal-bucketing/unit.date-time.truncate
+  "If the current breakout `unit` will not return at least [[temporal-filter-min-num-rows]], find the largest unit
+  that will."
+  [unit  :- ::lib.schema.temporal-bucketing/unit.date-time.truncate
+   start :- ::lib.schema.literal/temporal
+   end   :- ::lib.schema.literal/temporal]
+  (or
+   (let [num-points (shared.ut/unit-diff unit start end)]
+     (when (< num-points temporal-filter-min-num-rows)
+       (when-let [next-largest-unit (unit->next-largest-unit unit)]
+         (temporal-filter-find-best-breakout-unit next-largest-unit start end))))
+   unit))
+
+(mu/defn ^:private temporal-filter-update-breakouts :- ::lib.schema/query
+  "Update the first breakout against `column` so it uses `new-unit` rather than the original unit (if any); remove all
+  other breakouts against that column."
+  [query        :- ::lib.schema/query
+   stage-number :- :int
+   column       :- ::lib.schema.metadata/column
+   new-unit     :- ::lib.schema.temporal-bucketing/unit.date-time.truncate]
+  (transduce
+   identity
+   (fn
+     ([]
+      {:query query, :has-seen-column? false})
```

Comment:
Wouldn't it be clearer to use the 4 arity version of `transduce` instead? Or just use `reduce` and return `:query` from its result?

Reply:
fixed